csrc/scheduler/matmul_heuristic_plugin.cpp

```diff
@@ -0,0 +1,287 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+
+#include <ir/interface_nodes.h>
+#include <mma_type.h>
+#include <scheduler/matmul_heuristic_plugin.h>
+#include <scheduler/mma_utils.h>
+#include <utils.h>
+
+#include <dlfcn.h>
+#include <cstdint>
+
+namespace nvfuser {
+
+namespace matmul_heuristic_plugin {
+
+namespace {
+
+//! Defines HeuristicFuncPtr as type of the "getConfig" symbol
+typedef void (*HeuristicFuncPtr)(KernelConfig*, const ProblemDescription*);
+
+thread_local class PluginInterface {
```

Comment:
Does this mean we will initialize an object everytime we create a thread?

Reply:
Yes we will have one of these objects per thread, but they are pretty light-weight. Without thread_local I was hitting segfaults at dlclose when running pytest. The alternative is to guard the dlopen and dlclose parts with `std::lock_guard`; I'll try that out.

Reply:
Switched to keeping `plugin` global and using a mutex to guard whenever we modify it.